src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -702,14 +687,25 @@ private unsafe string[] GetValueNamesCore(int values)
                                 throw new IOException(SR.Arg_RegGetOverflowBug, e);
                             }
                         }
-                        char[] blob = new char[datasize / 2];
 
-                        ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blob, ref datasize);
+                        char[]? pooledArray = null;
+                        int blobLength = datasize / 2;
+                        Span<char> blob = blobLength <= 128 ? // arbitrary limit
+                            stackalloc char[blobLength] :
+                            (pooledArray = ArrayPool<char>.Shared.Rent(blobLength));
+
+                        fixed (char* blobPtr = &MemoryMarshal.GetReference(blob))
+                        {
+                            ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, blobPtr, ref datasize);
+                        }
 
                         // make sure the string is null terminated before processing the data
-                        if (blob.Length > 0 && blob[blob.Length - 1] != (char)0)
+                        if (blob.Length > 0 && blob[blobLength - 1] != '\0')
```

Comment:
If the registry value chanced between the two `RegQueryValueEx` calls, the blob can contain uninitialized data now (it is not zero-filled anymore) and this code can have unpredictable behavior.